### PR TITLE
WIP - .NET API client library first pass

### PR DESCRIPTION
**Pull request by jrowett**
not for merge - work in progress

### CODE REVIEW

File: src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    client.BaseAddress = endpoint;
+                    client.DefaultRequestHeaders.Accept.Clear();
+                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                    return client;
+                };
+            });
+
+            using var loggerFactory = LoggerFactory.Create(builder =>
+            {
+                builder.AddConsole();
```
Comment (vicr1) [inline src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs:35]: Is it within the remit of a client library to turn on console logging, or affect the logging configuration of the host application?

File: src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            });
+
+            var logger = loggerFactory.CreateLogger<NhsAppointmentsApi>();
+
+            services.AddSingleton<ILogger>(logger);
```
Comment (vicr1) [inline src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs:40]: Could this affect other logging configuration setup by the host application

**Review by vicr1 — CHANGES_REQUESTED** [review]

File: src/api/Nhs.Appointments.ApiClient/Impl/ApiClientBase.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            }        
+        }
+
+        private async void AssertResponseStatusCode(HttpResponseMessage response, params HttpStatusCode[] expected)
+        {
+            if(!expected.Contains(response.StatusCode))
+            {
+                var exception = new UnexpectedResponseException(response.StatusCode, expected);
+                var responseContent = await response.Content.ReadAsStringAsync();
+                logger.LogTrace(responseContent);
```
[the remark this answers is not in the document]
Comment (jrowett) [reply src/api/Nhs.Appointments.ApiClient/Impl/ApiClientBase.cs:86]: i'm relaxed about this but generally when using some other framework (whether ASP.Net, MassTransit, whatever) i appreciate getting useful console output from the start. MassTransit is a great example actually, it produces loads of telemetry. let me know if you want this removed though

File: src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs (hunk 1 above)
Comment (jrowett) [reply src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs:35]: fair point - i'm going to look more closely at how other OSS frameworks handle this

File: src/api/Nhs.Appointments.ApiClient/Models/AttendeeDetails.cs
[the remark this answers is not in the document]
Comment (jrowett) [reply src/api/Nhs.Appointments.ApiClient/Models/AttendeeDetails.cs:1]: happy with that approach - probably a candidate for a separate ticket

File: src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs (hunk 2 above)
Comment (jrowett) [reply src/api/Nhs.Appointments.ApiClient/Configuration/ContainerConfiguration.cs:40]: removed and updated in line with P&P guidance. library will emit logs if required but will not interfere with existing logging in the implementor's app

File: src/api/Nhs.Appointments.ApiClient/Impl/ApiClientBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrowett) [reply src/api/Nhs.Appointments.ApiClient/Impl/ApiClientBase.cs:86]: updated in line with guidance at https://learn.microsoft.com/en-us/dotnet/core/extensions/logging-library-authors

**Review by vicr1 — APPROVED** [review]

**Review by joefarina — APPROVED** [review]

**Pull request merged**